src/runtime/swg-client-event-manager.js

```diff
@@ -0,0 +1,312 @@
+    this.additionalParameters_ = value;
+  }
+
+  /**Returns true if the object is valid.  Otherwise it throws an error if you
+   * pass true and returns false if you don't.
+   * @param {boolean} createError
+   */
+  isValid(createError) {
+    if (!SwgClientEvent.isValidType_(this.eventType_)) {
+      if (!createError) {
+        return false;
+      }
+      throw new Error('Invalid event type in SwgClientEvent (' +
+          this.eventType_ + ')');
+    }
+    if (!SwgClientEvent.isValidOriginator_(this.eventOriginator_)) {
+      if (!createError) {
+        return false;
+      }
+      throw new Error('Invalid event originator in SwgClientEvent (' +
+          this.eventOriginator_ + ')');
+    }
+    if (!SwgClientEvent.isValidIsFromUserAction_(this.isFromUserAction_)) {
+      if (!createError) {
+        return false;
+      }
+      throw new Error('Invalid isFromUserAction in SwgClientEvent (' +
+          this.isFromUserAction_ + ')');
+    }
+    if (!SwgClientEvent.isValidAdditionalParameters_(
+        this.additionalParameters_)) {
+      if (!createError) {
+        return false;
+      }
+      throw new Error('Invalid additional parameters in SwgClientEvent (' +
+          this.additionalParameters_ + ')');
+    }
+    return true;
+  }
+}
+
+export class SwgClientEventManager {
+
+  /**True if the value is a function
+   * @param {*} value
+   * @private
+   */
+  static isFunction_(value) {
+    return typeof value === 'function';
+  }
+
+  constructor() {
+    /** @type {!CallBackArray} @private */
+    this.listeners_ = [];
+
+    /** @type {!CallBackArray} @private */
+    this.filterers_ = [];
+  }
+
+  /**Remove all existing filterers and listeners
+   */
+  clear() {
+    this.listeners_ = [];
+    this.filterers_ = [];
+  }
+
+  /**
+   * Ensures the callback function is notified anytime one of the passed
+   * events occurs unless a filterer returns false.
+   * @param {!SwgEventManagerListener} callback
+   * @public
+   */
+  addListener(callback) {
+    assert(SwgClientEventManager.isFunction_(callback),
+        'Invalid callback in SwgClientEventManager.addListener');
+    this.listeners_.push(callback);
+  }
+
+  /**
+   * Filterers are called before listeners, if it returns
+   * ShouldFilter.STOP_EXECUTING then the event is canceled.
+   * @param {!SwgEventManagerFilterer} callback
+   * @public
+   */
+  addFilterer(callback) {
+    assert(SwgClientEventManager.isFunction_(callback),
+        'Invalid callback in SwgClientEventManager.addFilterer');
+    this.filterers_.push(callback);
+  }
+
+  /**Call this function to inform all listeners that a client event has
+   * occurred.
+   * @param {!SwgClientEvent} event
+   * @public
+   */
+  logEvent(event) {
+    event.isValid(true);
+    let callbacks = this.filterers_;
+    for (let callbackNum = 0; callbackNum < callbacks.length; callbackNum++) {
+      if (callbacks[callbackNum](event) === ShouldFilter.STOP_EXECUTING) {
+        return false;
```

Comment:
does LogEvent has a return value? What is the caller expected to do with the return value?

Reply:
I think what you want to do is not log the event here? just return; should do

Reply:
I don't really know what the caller would do with this information but its very easy to inform them whether or not the event is filtered so why not?

Reply:
the callers are propensity APIs and analyticsService OSS logEvent API. These APIs don't expect a return value.